### ISSUE TITLE
Problem with private certificate

### ISSUE THREAD

**Issue by xlemassacre** [BimmerGestalt/HassGestalt#14]
Hi,

after logging in to my hass instance via HassGestalt, there is only a blank screen.
It seems that there is a problem with https and my custom certificate - when I connect via http it's working fine.

I use a custom CA with client authentication to connect to my hass instance. Would it be possible to support that in the app (with the Android OS built in certificate handling)?

**Comment by hufman** [comment]
It's not going anything special to configure the server cert, just passing the [websocket url directly](https://github.com/BimmerGestalt/HassGestalt/blob/7b05cd4d938309ddbfc487381fa8658bee2649dc/app/src/main/java/io/bimmergestalt/hassgestalt/hass/HassWsClient.kt#L15), which works well for my LetsEncrypt-provided cert. How is your CA registered on your phone? How do you have client authentication configured? Once the Oauth screen is clicked through (using your regular browser), HassGestalt needs to talk to the [Oauth Token](https://github.com/BimmerGestalt/HassGestalt/blob/7b05cd4d938309ddbfc487381fa8658bee2649dc/app/src/main/java/io/bimmergestalt/hassgestalt/OauthConstants.kt#L26) and [Websocket](https://github.com/BimmerGestalt/HassGestalt/blob/7b05cd4d938309ddbfc487381fa8658bee2649dc/app/src/main/java/io/bimmergestalt/hassgestalt/hass/HassWsClient.kt#L21) endpoints directly.

**Comment by xlemassacre** [comment]
I don't use lets encrypt here, but a self hosted certificate authority.

The home assistant server is running on kubernetes behind an ingress controller, with a server certifcate from that CA, and also mTLS enabled:
nginx.ingress.kubernetes.io/auth-tls-verify-client: "on"

On my phone (in android settings) I have the root certificate of the CA and also a client/user certificate installed.
This works for the browser and also for the home assistant app, but I guess this is not supported in HassGestalt at the moment.

I can reproduce it the following way:
-> open the HassGestalt app
-> enter url of my hass instance
-> forward to my hass instance for login
-> successfully login with my user
-> redirect to the HassGestalt app
-> no error but also no dashboards shown

**Comment by hufman** [comment]
I found [how the official app does it](https://github.com/home-assistant/android/blob/bf5a983be2e8a9db0a35251fbfa76ce59c7dddc9/common/src/main/java/io/homeassistant/companion/android/common/data/TLSHelper.kt), how does the UI flow work? The PR that added it said there's no GUI changes, but I am not sure how the KeyChainRepository would know what key and cert to use. Is there some sort of chooser dialog?

**Comment by xlemassacre** [comment]
Yes there is a chooser dialog at the first time where you need to select the certificate. This seems to be stored somehow, I don't have to select it every time.

**Comment by hufman** [comment]
The name of the key is [saved as a preference](https://github.com/home-assistant/android/blob/906e1f7a9bb52256fdcdb4d79017da5696aea69f/common/src/main/java/io/homeassistant/companion/android/common/data/keychain/KeyChainRepositoryImpl.kt), and then the key is loaded from [Android Keychain module](https://developer.android.com/reference/android/security/KeyChain) when needed.
